core/src/main/java/com/linecorp/armeria/common/stream/InputStreamStreamMessage.java

```diff
@@ -0,0 +1,384 @@
+        this.length = length;
+        return this;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return !completionFuture.isDone();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        if (isOpen()) {
+            return false;
+        }
+        final InputStreamSubscription inputStreamSubscription = this.inputStreamSubscription;
+        return inputStreamSubscription == null || !inputStreamSubscription.written;
+    }
+
+    @Override
+    public long demand() {
+        final InputStreamSubscription inputStreamSubscription = this.inputStreamSubscription;
+        if (inputStreamSubscription != null) {
+            return inputStreamSubscription.requested;
+        } else {
+            return 0;
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return completionFuture;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        if (!subscribedUpdater.compareAndSet(this, 0, 1)) {
+            subscriber.onSubscribe(NoopSubscription.get());
+            subscriber.onError(new IllegalStateException("Only single subscriber is allowed!"));
+            return;
+        }
+
+        if (executor.inEventLoop()) {
+            subscribe0(subscriber, executor, options);
+        } else {
+            executor.execute(() -> subscribe0(subscriber, executor, options));
+        }
+    }
+
+    private void subscribe0(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                            SubscriptionOption... options) {
+        final ExecutorService blockingTaskExecutor;
+        if (this.blockingTaskExecutor != null) {
+            blockingTaskExecutor = this.blockingTaskExecutor;
+        } else {
+            final ServiceRequestContext serviceRequestContext = ServiceRequestContext.currentOrNull();
+            if (serviceRequestContext != null) {
+                blockingTaskExecutor = serviceRequestContext.blockingTaskExecutor();
+            } else {
+                blockingTaskExecutor = CommonPools.blockingTaskExecutor();
+            }
+        }
+
+        final InputStreamSubscription inputStreamSubscription = new InputStreamSubscription(
+                subscriber, executor, blockingTaskExecutor, bufferSize, offset, length,
+                containsNotifyCancellation(options));
+        this.inputStreamSubscription = inputStreamSubscription;
+        subscriber.onSubscribe(inputStreamSubscription);
+
+        // To make sure to close the inputStreamSubscription when this is aborted.
+        if (completionFuture.isCompletedExceptionally()) {
+            completionFuture.exceptionally(cause -> {
+                inputStreamSubscription.close(cause);
```

Comment:
This callback will be executed in the event loop.
```suggestion
                inputStreamSubscription.close0(cause);
```